### PR TITLE
api/vmauth: introduce unauthorizedUserAccess field at spec

### PR DESCRIPTION
**Pull request by f41gh7**
Previously for unauthorized_user config section of vmauth. Operator used two configuration fields: `unauthorizedAccessConfig` and inlined fields from `VMUserOptions`. This behaviour doesn't aligh with configuration file supported
 by vmauth. It also incorrectly exposed fields from `VMUserOptions` at `spec`.
Which could mislead users, since `spec.default_urls` could be treated as global config option for vmauth, but in fact, it can only be used at `unauthorized_user` section.

 This commit replaces both fields with the new field `unauthorizedUserAccess`.
It combines both config options - `url_map` and `VMUserOptions`. Replaced fields marked as deprecated and will be removed at `v1.0` operator API release.

Also `VMauth` now properly validates `unauthorized_user` related configuration and returns proper error to the user, instead of crashing `vmauth` container in runtime.

Related issues:
- https://github.com/VictoriaMetrics/operator/issues/1168
- https://github.com/VictoriaMetrics/operator/issues/1169

### CODE REVIEW

File: api/operator/v1beta1/vmauth_types.go [hunk 1]
```diff
@@ -79,93 +98,177 @@ type VMAuthSpec struct {
...
+	urlPrefix, err := url.Parse(urlPrefixStr)
+	if err != nil {
+		return err
+	}
+	// Validate urlPrefix
+	if urlPrefix.Scheme != "http" && urlPrefix.Scheme != "https" {
+		return fmt.Errorf("unsupported scheme for `url_prefix: %q`: %q; must be `http` or `https`", urlPrefix, urlPrefix.Scheme)
+	}
+	if urlPrefix.Host == "" {
+		return fmt.Errorf("missing hostname in `url_prefix %q`", urlPrefix.Host)
```
Comment (Haleygo) [inline api/operator/v1beta1/vmauth_types.go]:
```suggestion
		return fmt.Errorf("missing hostname in `url_prefix %q`", urlPrefix)
```

File: api/operator/v1beta1/vmauth_webhook.go [hunk 2]
```diff
@@ -59,6 +62,24 @@ func (r *VMAuth) sanityCheck() error {
 				r.Spec.ExternalConfig.SecretRef.Name, r.Spec.ExternalConfig.SecretRef.Key)
 		}
 	}
+
+	if len(r.Spec.UnauthorizedAccessConfig) > 0 {
```
Comment (Haleygo) [inline api/operator/v1beta1/vmauth_webhook.go:68]: can also add a check for not having both `UnauthorizedAccessConfig` and `UnauthorizedUserAccess` specified.

File: api/operator/v1beta1/vmauth_types.go [hunk 3]
```diff
@@ -19,55 +20,73 @@ type VMAuthSpec struct {
...
+	// will be removed at v1.0 release
+	// +deprecated
+	// +kubebuilder:validation:Schemaless
+	// +kubebuilder:pruning:PreserveUnknownFields
+	UnauthorizedAccessConfig []UnauthorizedAccessConfigURLMap `json:"unauthorizedAccessConfig,omitempty" yaml:"unauthorizedAccessConfig,omitempty"`
+	// UnauthorizedUserAccess defines unauthorized_user config section of vmauth config
 	// +optional
-	UnauthorizedAccessConfig []UnauthorizedAccessConfigURLMap `json:"unauthorizedAccessConfig,omitempty"`
-	UserConfigOption         `json:",inline"`
+	UnauthorizedUserAccess *VMAuthUnauthorizedUserAccess `json:"unauthorizedUserAccess,omitempty" yaml:"unauthorizedUserAccess,omitempty"`
```
Comment (Haleygo) [inline api/operator/v1beta1/vmauth_types.go]: maybe name it `UnauthorizedUserAccessSpec` or `UnauthorizedUserAccessConfig`?

Comment (f41gh7) [reply api/operator/v1beta1/vmauth_types.go]: `UnauthorizedUserAccessSpec` looks better to me, thanks for suggestion!

File: api/operator/v1beta1/vmauth_types.go [hunk 4]
```diff
@@ -79,93 +98,177 @@ type VMAuthSpec struct {
...
+		}
+	}
+	if err := vmuua.VMUserConfigOptions.Validate(); err != nil {
+		return fmt.Errorf("incorrect UnauthorizedUserAccess options: %w", err)
+	}
+
+	return nil
+}
+
+// UnauthorizedAccessConfigURLMap defines
```
Comment (AndrewChubatiuk) [inline api/operator/v1beta1/vmauth_types.go]:
```suggestion
// UnauthorizedAccessConfigURLMap defines unauthorized_user section configuration for vmauth
```

**Review by AndrewChubatiuk — DISMISSED** [review]

**Pull request merged**